docs/NativeCodeGeneration.md

```diff
@@ -0,0 +1,141 @@
+} meta_data_table_t;
+
+meta_data_table_t xdp = { _get_programs, _get_maps, _get_helpers };
+```
+
+The meta-data table provides pointers to three functions that permit querying the list of programs, the list of maps, and the list of helper functions. The table name is derived from the ELF file name, with _ replacing any character that is not valid in a C variable name. This variable is the only globally visible variable in the generated C code.
+
+## Exported programs
+
+Each program in the generated C file is exported via a program_entry_t:
+```
+typedef struct _program_entry
+{
+    uint64_t (*function)(void*);
+    const char* name;
+} program_entry_t;
+```
+
+The skeleton framework can then use NMR to publish this information to the eBPF execution context.
+
+## Imported helper functions
+
+The generated C code exposes a table containing the address of each helper function, name, ID, and additional meta-data of the helper function. C code generator emits a table for the helper functions referenced by the program:
+
+```
+typedef struct _helper_function_entry
+{
+    uint64_t (*address)(uint64_t, uint64_t, uint64_t, uint64_t, uint64_t);
+    uint32_t helper_id;
+    const char* name;
+    bool tail_call;
+} helper_function_entry_t;
+
+static helper_function_entry_t _helpers[] = {
+    { NULL, 1, "helper_id_1"},
+    { NULL, 0, "helper_id_0"},
+};
+```
+
+The skeleton then uses NMR to query the address of each helper function in the table, after which it writes to the address field in the table. Calls to helper functions in the generated code are called indirectly via the address field.
+
+## Exported maps
+
+Each map referenced by any of the eBPF programs are added as a map_entry_t:
+
+```
+typedef struct _map_entry
+{
+    void* address;
+    ebpf_map_definition_in_file_t definition;
+    const char* name;
+} map_entry_t;
+
+static map_entry_t_maps[] = {
+    { NULL, { 36, 2, 4, 8, 1,  }, "dropped_packet_map" },
+    { NULL, { 36, 2, 4, 4, 1,  }, "interface_index_map" },
+};
+```
+
+The skeleton then uses NMR to either create an instance of the map or obtain the address of an existing map and then write the address into the address field. References to the maps in the generated code are then indirect via the address field.
+
+## Loading an EBPF program from a PE .sys file
+
+The process of loading an eBPF program is a series of interactions between the eBPF Execution Context and the generated .sys file.
+
+### Lifecycle of a generaed PE .sys file
+
+1) During driver entry, the generated DriverEntry will call [NmrRegisterClient](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/netioddk/nf-netioddk-nmrregisterclient) for the helper contract and the map contract.
+
+2) In response to the attach notification on the helper function NMR contract, the skeleton will query the address of each helper function the eBPF program used and write the address into the _helpers table.
```

Comment:
Do you have to do anything special to make tail calls work?  (I don't think so, but want to know if you've verified that.)

Reply:
This is part of the helper function metadata that is returned from the EC. I added clarifying text.